File: problems/LinkedListCycle.java


public class LinkedListCycle {

    public static class LinkedListNode {
        public int value;
        public LinkedListNode next;

        public LinkedListNode(int value) {
            this.value = value;
        }
    }

    private static boolean containsCycle(LinkedListNode head) {
        LinkedListNode leader = head;
        LinkedListNode chaser = head;
        if (leader.next == null)
            return false;

        while (leader.next != null && leader.next.next != null) {
            leader = leader.next.next;
            chaser = chaser.next;
            if (leader == chaser) {
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        LinkedListNode a = new LinkedListNode(11);
        LinkedListNode b = new LinkedListNode(12);
        LinkedListNode c = new LinkedListNode(13);
        LinkedListNode d = new LinkedListNode(14);
        LinkedListNode e = new LinkedListNode(15);
        LinkedListNode f = new LinkedListNode(16);

        a.next = b;
        b.next = c;
        c.next = d;
        d.next = e;
        e.next = c;

        LinkedListNode one = new LinkedListNode(1);
        LinkedListNode two = new LinkedListNode(2);
        LinkedListNode three = new LinkedListNode(3);
        LinkedListNode four = new LinkedListNode(4);
        LinkedListNode five = new LinkedListNode(5);
        LinkedListNode six = new LinkedListNode(6);

        one.next = two;
        two.next = three;
        three.next = four;
        four.next = five;
        five.next = six;

        boolean testOne = containsCycle(a);
        boolean testTwo = containsCycle(one);

        System.out.println(testOne);
        System.out.println(testTwo);
    }
}

File: problems/python/linked_list_cycle.py


# Bonus
# 1) How would you detect the first node in the cycle? Define the first
# node of the cycle as the one closest to the head of the list.
# 2) Would the program always work if the fast runner moves three steps
# every time the slow runner moves one step?
# 3) What if instead of a simple linked list, you had a structure where 
# each node could have several "next" nodes? This data structure is
# called a "directed graph." How would you test if your directed
# graph had a cycle?


class LinkedListNode:

    def __init__(self, value):
        self.value = value
        self.next  = None

def contains_cycle(head):
    fast = head
    slow = head.next
    while fast is not None and fast.next is not None:
        if fast == slow:
            return True
        fast = fast.next.next
        slow = slow.next
    return False

def main():
    pass


if __name__ == '__main__':
    main()
